### ISSUE TITLE
Boost 1.64 support 

### ISSUE THREAD

**Issue by guwirth** [etas/vs-boost-unit-test-adapter#197]
* support boost 1.64
* test with boost 1.64
* ensure compatibility 

http://www.boost.org/doc/libs/1_64_0_b2/libs/test/doc/html/boost_test/change_log.html

**Comment by netspiri** [comment]
Currently busy with other projects and cannot proceed on this but from what I gleamed from the release notes, I am not sure the adapter will work with modules compiled with Boost.Test for Boost 1.64.

Reason being pull [108](https://github.com/boostorg/test/pull/108/files) changed key symbol names which we depend upon. Since we utilise the debug symbols as a first measure to identify Boost.Test module (and to avoid execution of possibly malicious modules), this will not work anymore.

This is a listing of the debug symbol names which we have used till now, and will need to test for in future releases:
```
"boost::unit_test::runtime_config::list_content" 	// Boost 1.59
"boost::unit_test::runtime_config::LIST_CONTENT" 	// Boost 1.60 - Boost 1.63
"boost::unit_test::runtime_config::list_content_str" 	// Boost 1.64

"boost::unit_test::runtime_config::VERSION" 		// Boost 1.63
"boost::unit_test::runtime_config::version_str" 	// Boost 1.64
```

**Comment by abernosk** [comment]
Hi there, I'm wondering if this issue relates to a problem I am having after upgrading to boost 1.64 in my app that is built with Visual Studio 2015 R3. Visual Studio is unable to find any boost unit tests resulting in the Visual Studio's Test Explorer window being empty. In the output window after successful build and during test discovery I get "========== Discover test finished: 0 found". I also don't see any tests when I run the simple  "Boost Unit Test Project" and follow the instructions from the Quick Start guide in the documentation. Everything worked as documented with Boost 1.60 but now I need 1.64 so can you tell me if VS 2015 R3 + latest vs-boost-unit-test-adapter are compatible? I also went thru the troubleshooting guide and everything on my system appears to be healthy. Thanks in advance. -Adam

**Comment by guwirth** [comment]
> Since we utilise the debug symbols as a first measure to identify Boost.Test module (and to avoid execution of possibly malicious modules), this will not work anymore.

Problem seems to be that with every new boost version these symbol (names) are changing.
* Just an idea: Why not adding an own stable symbol to the binary?
* or remove this automatic Boost.Test detection and allow only fixed EXE extensions/namings?
* anyway the question how big performance impact of reading symbol file is?

@netspiri do you know how Google detects that the binary contains Google.Test code?

Reading symbols is at: https://github.com/etas/vs-boost-unit-test-adapter/blob/99a8bacc3fc58203819d57327883ba8df44bedfe/BoostTestAdapter/Boost/Runner/BoostTestRunnerBase.cs#L67

**Comment by netspiri** [comment]
@guwirth 

> Just an idea: Why not adding an own stable symbol to the binary?
This is highly intrusive and we should not muck about with the generated executable. Even still, solely from an adapter point-of-view this is out of our abilities. We would need to resort to adapting the project template and still, at that point, we would not allow support for test modules which are built without the use of said template.

> or remove this automatic Boost.Test detection and allow only fixed EXE extensions/namings?
Already available by using `.test.boostd.exe`. I would suggest against from removing it entirely since this creates friction - test adapter users should be able to use the adapter with minimal interference. If we force the user to change the test module or project settings just for the sake of the test adapter, I would say that we would have failed at providing a good user experience.

> anyway the question how big performance impact of reading symbol file is?
Will get back to you on this. In the process of adding new timing logs.

> @netspiri do you know how Google detects that the binary contains Google.Test code?
I am not familiar with the code of that particular adapter and given the little exposition I have with that code base, I would not like to jump to incorrect conclusions. Maybe @csoltenborn might be of help.

**Comment by guwirth** [comment]
>> @netspiri do you know how Google detects that the binary contains Google.Test code?

> I am not familiar with the code of that particular adapter and given the little exposition I have with that code base, I would not like to jump to incorrect conclusions. Maybe @csoltenborn might be of help.

@LukaszMendakiewicz do you know how the Google adapter detects Google.Test binaries?

**Comment by guwirth** [comment]
>> anyway the question how big performance impact of reading symbol file is?

> Will get back to you on this. In the process of adding new timing logs.

@netspiri isn't it possible to attach with a second VS instance to a running Boost.Test extension with a profiler?

**Comment by LukaszMendakiewicz** [comment]
We addressed this in our fork by scanning for both symbols, `LIST_CONTENT` for Boost.Test < 1.64 and `btrt_list_content` for Boost.Test >= 1.64.
See:
https://github.com/Microsoft/vs-boost-unit-test-adapter/blob/a82fd1e1d8b5d3baedea76397231959e615e18c3/BoostTestAdapter/Boost/Runner/BoostTestRunnerBase.cs#L81

That's definitely not ideal, as it will do the scan multiple times, but the hope is that Boost.Test will settle on one verbiage and we will keep only that one as support for older Boost.Test versions is phased out.

It is addressed in Google Test adapter similarly, looking for a tell-tale in the executable. The difference is theirs solution is scanning for specific strings directly rather than symbols through DbgHelp. See:
https://github.com/csoltenborn/GoogleTestAdapter/commit/bb018b761194375b9e45fcf6447ce6ba2f84f9b6#diff-45fea78ea18db8b014b829961b0ea61aR29
Note that it is relatively fresh code, which I do not think have shipped in any release yet.

There is also a fast-path, similar to yours `.test.boostd.exe`, checking for existence of `{executable-name}.is_google_test` file.

**Comment by guwirth** [comment]
@netspiri and @LukaszMendakiewicz 

Looking to [Google.Test](https://github.com/csoltenborn/GoogleTestAdapter/commit/bb018b761194375b9e45fcf6447ce6ba2f84f9b6#diff-45fea78ea18db8b014b829961b0ea61aR29) think they have the better and more save approach. They open the binary as file and search with Boyer-Moore algorithm for strings in the binary. Think this is better
* because typically the EXE is much smaller as the symbol file
* it's also possible to search for substrings like `boost::unit_test`
* looking for `boost::unit_test` is more generic and working with all boost versions
* it's possible to search in debug and release version also without creating debug information

It tried it with the SysInternal tool `strings`. The Boost.Test binaries all contain the ASCII string `boost::unit_test` at least once.

So proposal is to change this here and in the Microsoft fork.

**Comment by csoltenborn** [comment]
Being a bit late to the party, here are two things to consider:
* Currently my code reads the complete file with `File.ReadAllBytes(executable)`. I guess that this should be appropriate for executables, but one could of course think about reading the file in chunks especially if...
* ... there are sections of the file which contain (or are at least more likely to contain) the strings (I have no knowledge in that area)

**Comment by guwirth** [comment]
@csoltenborn thx for your feedback.

* Looking to our projects the binaries are about 500k to 80MB. Compared to the symbol files with 30MB to 500MB they are really "small". This 80MB EXEs and 500MB symbol files are from extensive static linking.
* Not sure if a "better algorithm" is really faster. Only possibility I know is to read the PE header ([Portable Executable File Format](https://msdn.microsoft.com/en-us/library/ms809762.aspx)). Because you have to read and start the binary later one anyway I think the OS cache will do here a good job (to be proved).
* Having a look with a PE viewer (like pestudio):
  * symbols are not part of imports/exports section (because we are linking statically)
  * symbols are not part of resources
  * they are part of .text segments

==> So reading all .text segments and searching only there could be a solution?

**Comment by csoltenborn** [comment]
I'm also not sure at all whether such optimizations are worth the effort, since the time for investigating an executable is heavily dominated by loading that executable into memory, and `File.ReadAllBytes()` seems to be very fast with respect to that. Plus, users might have configured a different compiler (e.g. clang), which might influence where to look for the strings.

I guess that I will stay with reading the complete file for now (I'm going to release this in about three weeks after having returned from my well-deserved vacation), and only react if there are problems - and if this is the case, we could try to come up with a smarter way (and share that code).

**Comment by guwirth** [comment]
https://github.com/etas/vs-boost-unit-test-adapter/wiki/User-Manual#force-list-content-discovery

* Should also debug/release be supported?
  * debug : *.test.boostd.exe
  * release: *.test.boost.exe

Or is ***.test.boost.exe** the stable extension and the **D** should added in the prior filename part?
* Sample.test.boost**d**.exe
* Sample.test.boost.exe
-or-
* Sample**D**.test.boost.exe
* Sample.test.boost.exe

@LukaszMendakiewicz & @csoltenborn how is Google.Test extension doing this?

**Comment by LukaszMendakiewicz** [comment]
GTA/TAfGT are not checking the executable file name. There is a check for a separate file, `{executable-name}.is_google_test`, e.g. `Sample.exe.is_google_test` for `Sample.exe`.

**Comment by csoltenborn** [comment]
In fact, I do check for file names, but only as an alternative (and for backwards compatibility): users can provide a regex via the adapter's settings. Here's the [according code](https://github.com/csoltenborn/GoogleTestAdapter/blob/092ffbb4753f00bd36aecab997406150a20a12b1/GoogleTestAdapter/Core/GoogleTestDiscoverer.cs#L91)...

Note that this code is not yet released (as Lukasz has already mentioned above).

**Comment by guwirth** [comment]
> In fact, I do check for file names, but only as an alternative (and for backwards compatibility): users can provide a regex via the adapter's settings. Here's the according code... Note that this code is not yet released (as Lukasz has already mentioned above).

* I like the solution checking the filename only. This is for sure much faster as parsing the binary.
* I'm not a big fan of wasting the repository with additional files. So my favourite is still a predefined file extension.
* Think consequent would be to use `*.boost.test.exe` as extension (not `*.test.boostd.exe`) and put configuration specific parts into the part before this:
  * Sample**D**.boost.test.exe
  * Sample.boost.test.exe

> GTA/TAfGT are not checking the executable file name. There is a check for a separate file, {executable-name}.is_google_test, e.g. Sample.exe.is_google_test for Sample.exe.

Here you have the same problem that you need two files: one for the debug and one for the release version (in case they have different filenames).

Means should try in this order:
1. if a `<Filters>`is defined use this / if not `*.boost.test.exe` is the default
1. search binary with filter
1. if can't find a file with filter, use Google.Test algorithm to find string in binary

**Comment by csoltenborn** [comment]
I think we will stay as-is for the following reasons:
* binary search is rather fast and shouldn't bother our users
* they neither have to stick to some naming scheme, nor have to provide any settings - it just works out of the box
* finally, we want to avoid running non-gtest executables by all means. In our case, this happens during test discovery, i.e., running the executables is triggered just by building the solution. If somebody has that `FormatTest.exe` which formats `C:` (but is not a gtest executable), we do not want to run that executable. This is probably not very likely, but still...

The `is_google_test` file as well as the regular expression are only supposed to cover very weird cases.

**Comment by guwirth** [comment]
solved in 1.0.9.32:
https://ci.appveyor.com/api/buildjobs/i7hw6uevj5la5l94/artifacts/BoostTestPlugin%2Fbin%2FRelease%2FBoostUnitTestAdapter.vsix
